File: app/src/main/java/com/github/ykiselev/playground/init/FrameTask.java
package com.github.ykiselev.playground.init;

public abstract class FrameTask implements Runnable {

    private final Runnable next;

    public FrameTask(Runnable next) {
        this.next = next;
    }

    protected abstract void onFrameStart();

    protected abstract void onFrameEnd();

    @Override
    public void run() {
        onFrameStart();
        if (next != null) {
            next.run();
        }
        onFrameEnd();
    }
}

File: app/src/main/kotlin/com/github/ykiselev/playground/init/FrameTask.kt
package com.github.ykiselev.playground.init

abstract class FrameTask(private val next: Runnable?) : Runnable {

    protected open fun onFrameStart() {}

    protected open fun onFrameEnd() {}

    override fun run() {
        onFrameStart()
        next?.run()
        onFrameEnd()
    }
}
